### PR TITLE
[TypeDeclaration] Skip ReturnTypeDeclarationRector when class parent extends is in 'vendor'

### LINKED ISSUE #5205: Incorrect behavior of ReturnTypeDeclarationRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | 0.8.56, demo ver ?     |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/9cba5a13-fa55-4f06-880a-ccdd8259db78

```php
<?php

class DemoFile
{
    public function run()
    {
        throw new Exception("not implemented");
    }
}

class DemoFile2 extends DemoFile {
    
    public function run() : Baba {
       return new Baba();        
    }
}

class Baba {
}
```

Rector diff:
```patch
 
 class DemoFile2 extends DemoFile {
     
-    public function run() : Baba {
+    public function run() : void {
        return new Baba();        
     }
 }
 
 class Baba {
}
```
### Responsible rules

* `ReturnTypeDeclarationRector`

## Expected Behavior

Should be skipped.

In my own code it has a little bit different behaviour for some reason: it does add `: void` to the throwing function as a return type, which makes the code screaming about compatibility between the overriden function.

### PR DESCRIPTION
**Pull request by samsonasik**
Closes #5210 Fixes #5205

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/type-declaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php [hunk 1]
```diff
@@ -248,6 +256,27 @@ private function addReturnType(FunctionLike $functionLike, Node $inferredReturnN
...
+
+        $parentName = $parent->extends->toString();
+        $reflectionClass = new ReflectionClass($parentName);
+        if ($reflectionClass->isInternal()) {
+            return true;
+        }
+
+        $fileName = $reflectionClass->getFileName();
+        return Strings::contains((string) $fileName, 'vendor');
+    }
```
Comment (TomasVotruba) [inline rules/type-declaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php:278]: This logic should be extracted to `packages/vendor-locker`

There might be some useful class like `Rector\VendorLocker\NodeVendorLocker\ClassMethodVendorLockResolver`

File: phpstan.neon [hunk 2]
```diff
@@ -577,3 +577,7 @@ parameters:
 #            paths:
                 # weird case
 #                - packages/better-php-doc-parser/src/PhpDocParser/AnnotationContentResolver.php
+
+        -
+            message: '#Class cognitive complexity is 32, keep it under 30#'
+            path: rules/type-declaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php
```
Comment (TomasVotruba) [inline phpstan.neon:583]: Crossing cognitive complexity treshold should not be ignored.
Instead, should be solved by a new service extraction.

It works well with duplicated method PHPStan rule. That way we keep small consistent services with specific work and without duplication.

File: rules/type-declaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php (hunk 1 above)
Comment (samsonasik) [reply rules/type-declaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php]: implemented with use `VendorLockResolver->isReturnChangeVendorLockedIn()`

File: phpstan.neon (hunk 2 above)
Comment (samsonasik) [reply phpstan.neon]: fixed.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Comment by TomasVotruba** [comment]
@samsonasik There is one pull-request that is related to this issue.

Could you verify it was fixed or handle it?
https://github.com/rectorphp/rector/pull/5209

**Comment by samsonasik** [comment]
I will try.

**Pull request merged**